File: ios-app/Stripe/Simulation.swift
import Foundation

/// I pochi numeri che imposti tu. Tutto il resto della dashboard, fino
/// all'ultima cifra del grafico, viene calcolato da qui: cosi' non esistono
/// due valori che possono contraddirsi.
struct Simulation: Codable, Equatable {
    var merchantName = "Digital Consult LLC"
    var currency = Currency.usd

    /// Quanto entra in un giorno. Ogni giornata pesca un importo a caso in
    /// questo intervallo.
    var dailyMin: Double = 400
    var dailyMax: Double = 800

    /// Gli importi dei singoli pagamenti, gli stessi delle notifiche. La
    /// giornata si riempie sommando questi finche' non arriva al suo totale:
    /// e' anche il motivo per cui il numero di pagamenti torna sempre.
    var paymentAmounts: [Double] = [9.99, 49.99]

    /// Quanto scende il netto rispetto al lordo.
    var netDeductionPercent: Double = 2

    /// Che quota di clienti, ogni giorno, prende anche il prodotto piu' caro
    /// dopo quello base. Non e' un numero fisso: ogni giornata pesca il suo in
    /// questo intervallo, altrimenti pagamenti e clienti resterebbero sempre
    /// nella stessa proporzione e si vedrebbe che sono inventati.
    ///
    /// Opzionali per non rompere i salvataggi fatti prima.
    var repeatMinPercent: Double?
    var repeatMaxPercent: Double?

    var upsellRange: (low: Double, high: Double) {
        let low = repeatMinPercent ?? 40
        let high = repeatMaxPercent ?? 80
        return (min(low, high), max(low, high))
    }

    /// Incassi fissati a mano, per data di calendario (`AAAA-MM-GG`). La
    /// chiave e' la data e non "quanti giorni fa" proprio perche' quello che
    /// fissi oggi deve restare su oggi anche domani. Il totale esatto dipende
    /// dai tagli disponibili: la giornata si riempie finche' ci sta, quindi
    /// arriva appena sotto.
    var dayTargets: [String: Double]?

    /// Da quanti giorni l'attivita' e' aperta: serve solo alla voce ALL.
    var businessDays: Int = 400

    /// Cambiandolo cambiano tutti i numeri, restando dentro gli stessi
    /// intervalli. E' quello che rende la simulazione stabile: senza, i valori
    /// ballerebbero a ogni ridisegno della schermata.
    var seed: UInt64 = 20_260_802

    enum Currency: String, Codable, CaseIterable, Identifiable {
        case usd, eur, gbp, aed

        var id: String { rawValue }

        /// Stripe scrive "US$" nella fascia in alto e "$" nei report. Non e'
        /// una svista nostra: e' cosi' nell'app vera.
        var cardSymbol: String {
            switch self {
            case .usd: return "US$"
            case .eur: return "€"
            case .gbp: return "£"
            case .aed: return "AED "
            }
        }

        var reportSymbol: String {
            switch self {
            case .usd: return "$"
            case .eur: return "€"
            case .gbp: return "£"
            // Il dirham non ha un simbolo che Stripe usi: scrive la sigla.
            case .aed: return "AED "
            }
        }

        var title: String {
            switch self {
            case .usd: return "Dollar"
            case .eur: return "Euro"
            case .gbp: return "Pound"
            case .aed: return "Dirham (AED)"
            }
        }
    }
}

extension Simulation {
    /// Qualche nome gia' pronto, per non doverlo inventare ogni volta. Sono
    /// inventati: non corrispondono ad aziende esistenti.
    static let suggestedNames = [
        "Tech Digital Hub",
        "Northgate Web Studio",
        "Vertex Media Group",
        "Lumen Digital Agency",
        "Ardent Web Consulting",
    ]
}

/// Una giornata: i singoli pagamenti incassati. Tutto il resto sono conti su
/// questa lista.
struct SimulatedDay {
    let offset: Int
    let payments: [Double]

    /// Quanti compratori distinti. Meno dei pagamenti, perche' chi fa upsell
    /// paga due volte.
    let customers: Int

    var gross: Double { payments.reduce(0, +) }
    var paymentCount: Int { payments.count }
}

// MARK: - Date

extension Simulation {
    /// Giorni dal 1970-01-01 alla data di calendario. E' l'algoritmo civile di
    /// Hinnant: solo interi, nessun fuso orario di mezzo, cosi' Swift e
    /// TypeScript danno per forza lo stesso numero.
    static func daysFromCivil(_ y: Int, _ m: Int, _ d: Int) -> Int {
        let anno = y - (m <= 2 ? 1 : 0)
        let era = Int(floor(Double(anno) / 400))
        let yoe = anno - era * 400
        let doy = (153 * (m + (m > 2 ? -3 : 9)) + 2) / 5 + d - 1
        let doe = yoe * 365 + yoe / 4 - yoe / 100 + doy
        return era * 146_097 + doe - 719_468
    }

    /// La data di `offset` giorni fa, secondo il calendario locale.
    static func date(at offset: Int, now: Date = Date()) -> Date {
        let cal = Calendar.current
        return cal.date(byAdding: .day, value: -offset, to: cal.startOfDay(for: now)) ?? now
    }

    /// La chiave con cui si fissa un incasso: `AAAA-MM-GG`.
    static func dateKey(_ date: Date) -> String {
        let c = Calendar.current.dateComponents([.year, .month, .day], from: date)
        return String(format: "%04d-%02d-%02d", c.year ?? 0, c.month ?? 0, c.day ?? 0)
    }

    /// Il numero di giorno assoluto usato come seme. Prima si usava `offset`,
    /// cioe' quanti giorni fa: a mezzanotte ogni giornata scivolava di un posto
    /// e tutto lo storico si rigenerava. Legandolo alla data, una giornata vale
    /// sempre lo stesso importo.
    static func epochDay(_ offset: Int, now: Date = Date()) -> Int {
        let c = Calendar.current.dateComponents([.year, .month, .day], from: date(at: offset, now: now))
        return daysFromCivil(c.year ?? 1970, c.month ?? 1, c.day ?? 1)
    }
}

// MARK: - Generazione

extension Simulation {
    /// La giornata a `offset` giorni fa: 0 e' oggi, 1 ieri.
    ///
    /// Il risultato dipende da seed e data, mai dall'ordine in cui la si
    /// chiama: domani questa stessa giornata varra' ancora quanto vale adesso. Senza questa garanzia gli incassi cambierebbero mentre
    /// scorri, e i totali non tornerebbero mai con il grafico.
    /// La giornata si costruisce per clienti, non per pagamenti sciolti: ognuno
    /// prende il prodotto base, e una parte aggiunge quello piu' caro. E' un
    /// upsell, non un riacquisto dello stesso taglio, ed e' il motivo per cui i
    /// pagamenti sono sempre piu' dei clienti.
    func day(_ offset: Int) -> SimulatedDay {
        let amounts = paymentAmounts.filter { $0 > 0 }.sorted()
        guard let base = amounts.first else {
            return SimulatedDay(offset: offset, payments: [], customers: 0)
        }
        let upsells = Array(amounts.dropFirst())

        let giorno = Simulation.epochDay(offset)
        var rng = SeededRandom(seed &+ UInt64(bitPattern: Int64(giorno) &* 7_919))
        let low = min(dailyMin, dailyMax)
        let high = max(dailyMin, dailyMax)

        // Ogni blocco di trenta giorni ha un suo livello medio, e la giornata
        // oscilla intorno a quello. Pescando ogni giorno in modo indipendente
        // il grafico veniva una linea piatta: mesi buoni e mesi scarsi sono
        // quello che dà alla curva la forma che ha nell'originale. Il valore
        // resta comunque dentro l'intervallo impostato.
        let block = Int(floor(Double(giorno) / 30))
        var blockRng = SeededRandom(seed &+ UInt64(bitPattern: Int64(block) &* 2_654_435_761))
        let centre = 0.25 + 0.5 * blockRng.double()
        let spread = (rng.double() - 0.5) * 0.4

        // Le estrazioni sopra avvengono comunque, anche quando oggi e' fissato:
        // saltarle sposterebbe tutto il flusso del generatore e cambierebbe le
        // giornate successive.
        let casuale = low + min(1, max(0, centre + spread)) * (high - low)
        let fissato = dayTargets?[Simulation.dateKey(Simulation.date(at: offset))]
        let target = (fissato ?? 0) > 0 ? fissato! : casuale

        // La quota di upsell del giorno, dal suo generatore: cosi' il flusso
        // principale resta identico fra Swift e TypeScript.
        var upsellRng = SeededRandom(seed &+ UInt64(bitPattern: Int64(giorno) &* 15_485_863))
        let range = upsellRange
        let quota = range.low + upsellRng.double() * (range.high - range.low)

        var payments: [Double] = []
        var customers = 0
        var sum = 0.0

        // Il tetto evita che un importo minuscolo con un incasso alto generi
        // decine di migliaia di righe.
        while sum + base <= target && payments.count < 2_000 {
            payments.append(base)
            sum += base
            customers += 1

            guard !upsells.isEmpty, rng.double() * 100 < quota else { continue }
            let extra = upsells[rng.index(upsells.count)]
            if sum + extra <= target {
                payments.append(extra)
                sum += extra
            }
        }

        return SimulatedDay(offset: offset, payments: payments, customers: customers)
    }

    func customerCount(_ day: SimulatedDay) -> Int { day.customers }

    /// Quanto e' entrato dall'inizio della giornata fino a `upToHour`, ora per
    /// ora. I pagamenti si distribuiscono fra le 6 e le 23, con piu' peso nel
    /// pomeriggio. Deve restare identica a `cumulativeByHour` in
    /// `src/lib/simulation.ts`: la cifra in alto viene da qui.
    func cumulativeByHour(_ offset: Int, upToHour: Int) -> [Double] {
        let d = day(offset)
        var rng = SeededRandom(seed &+ UInt64(bitPattern: Int64(Simulation.epochDay(offset)) &* 104_729))

        var hours = [Double](repeating: 0, count: 24)
        for amount in d.payments {
            let hour = min(23, Int(6 + rng.double() * 18))
            hours[hour] += amount
        }

        var out: [Double] = []
        var running = 0.0
        for hour in 0...min(23, max(0, upToHour)) {
            running += hours[hour]
            out.append(running)
        }
        return out
    }

    /// Quanto e' entrato finora oggi: l'ultimo punto della curva delle ore.
    func grossSoFar(now: Date = Date(), calendar: Calendar = .current) -> Double {
        cumulativeByHour(0, upToHour: calendar.component(.hour, from: now)).last ?? 0
    }

    func net(_ gross: Double) -> Double {
        gross * (1 - netDeductionPercent / 100)
    }
}

/// Generatore deterministico (SplitMix64). Non usa `Double.random`, che
/// darebbe numeri diversi a ogni chiamata: qui gli stessi ingressi devono
/// dare sempre le stesse cifre.
struct SeededRandom {
    private var state: UInt64

    init(_ seed: UInt64) { state = seed }

    mutating func next() -> UInt64 {
        state = state &+ 0x9E37_79B9_7F4A_7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58_476D_1CE4_E5B9
        z = (z ^ (z >> 27)) &* 0x94D0_49BB_1331_11EB
        return z ^ (z >> 31)
    }

    /// Un numero fra 0 e 1.
    mutating func double() -> Double {
        Double(next() >> 11) / Double(1 << 53)
    }

    /// Un indice valido per una lista di `count` elementi.
    mutating func index(_ count: Int) -> Int {
        count <= 1 ? 0 : min(count - 1, Int(double() * Double(count)))
    }
}

File: src/lib/simulation.ts
/**
 * La stessa simulazione della app iOS, riscritta qui.
 *
 * Non è una riscrittura "equivalente": deve produrre le identiche cifre, fino
 * al centesimo, perché telefono e browser leggono le stesse impostazioni da
 * Redis e devono mostrare gli stessi numeri. Per questo il generatore è
 * SplitMix64 su BigInt e non `Math.random`, e per questo l'ordine delle
 * operazioni ricalca riga per riga `ios-app/Stripe/Simulation.swift`.
 *
 * Se tocchi l'algoritmo qui, va toccato anche lì.
 */

const MASK = (1n << 64n) - 1n;

export type Currency = 'usd' | 'eur' | 'gbp' | 'aed';

export interface Simulation {
  merchantName: string;
  currency: Currency;
  dailyMin: number;
  dailyMax: number;
  paymentAmounts: number[];
  netDeductionPercent: number;
  /** Che quota di clienti, ogni giorno, prende anche il prodotto piu' caro
   *  dopo quello base. Opzionali per non rompere i salvataggi vecchi. */
  repeatMinPercent?: number;
  repeatMaxPercent?: number;
  /** Incassi fissati a mano, per data di calendario (`AAAA-MM-GG`). Quello che
   *  fissi oggi resta su oggi anche domani: e' la ragione per cui la chiave e'
   *  la data e non "quanti giorni fa". */
  dayTargets?: Record<string, number>;
  businessDays: number;
  /** Il giorno assoluto in cui l'attivita' ha aperto: da li' parte la rampa di
   *  crescita. Fisso, altrimenti col passare del tempo ogni giornata
   *  scivolerebbe indietro sulla rampa e cambierebbe valore. */
  startDay?: number;
  /** Resta sotto 2^53: oltre, JSON e JavaScript lo arrotonderebbero e il
   *  telefono e il browser genererebbero numeri diversi. */
  seed: number;
}

export const DEFAULT_SIMULATION: Simulation = {
  merchantName: 'Digital Consult LLC',
  currency: 'usd',
  dailyMin: 400,
  dailyMax: 800,
  paymentAmounts: [9.99, 49.99],
  netDeductionPercent: 2,
  repeatMinPercent: 40,
  repeatMaxPercent: 80,
  businessDays: 400,
  seed: 20260802,
};

export const SUGGESTED_NAMES = [
  'Tech Digital Hub',
  'Northgate Web Studio',
  'Vertex Media Group',
  'Lumen Digital Agency',
  'Ardent Web Consulting',
];

/** Quote fisse sul venduto: rimborsi, bloccati e falliti. Non sono impostabili
 *  perché la app iOS non le mostra, e un campo che esiste solo di qua si
 *  perderebbe al primo salvataggio dal telefono. */
const REFUNDED_RATIO = 0.031;
const BLOCKED_RATIO = 0.025;
const FAILED_RATIO = 0.175;

/** Giorni dal 1970-01-01 alla data di calendario. E' l'algoritmo civile di
 *  Hinnant: solo interi, nessun fuso orario di mezzo, cosi' Swift e TypeScript
 *  danno per forza lo stesso numero. */
export function daysFromCivil(y: number, m: number, d: number): number {
  const anno = y - (m <= 2 ? 1 : 0);
  const era = Math.floor(anno / 400);
  const yoe = anno - era * 400;
  const doy = Math.floor((153 * (m + (m > 2 ? -3 : 9)) + 2) / 5) + d - 1;
  const doe = yoe * 365 + Math.floor(yoe / 4) - Math.floor(yoe / 100) + doy;
  return era * 146097 + doe - 719468;
}

/** La data di `offset` giorni fa, secondo il calendario locale. */
export function dateAt(offset: number, now = new Date()): Date {
  return new Date(now.getFullYear(), now.getMonth(), now.getDate() - offset);
}

/** La chiave con cui si fissa un incasso: `AAAA-MM-GG`. */
export function dateKey(date: Date): string {
  const p = (n: number) => String(n).padStart(2, '0');
  return `${date.getFullYear()}-${p(date.getMonth() + 1)}-${p(date.getDate())}`;
}

/** Il numero di giorno assoluto usato come seme. Prima si usava `offset`, cioe'
 *  quanti giorni fa: a mezzanotte ogni giornata scivolava di un posto e tutto
 *  lo storico si rigenerava. Legandolo alla data, una giornata vale sempre lo
 *  stesso importo. */
export function epochDay(offset: number, now = new Date()): number {
  const d = dateAt(offset, now);
  return daysFromCivil(d.getFullYear(), d.getMonth() + 1, d.getDate());
}

// MARK: - Generatore

/** SplitMix64, identico a quello Swift. */
class SeededRandom {
  private state: bigint;

  constructor(seed: bigint) {
    this.state = seed & MASK;
  }

  next(): bigint {
    this.state = (this.state + 0x9e3779b97f4a7c15n) & MASK;
    let z = this.state;
    z = ((z ^ (z >> 30n)) * 0xbf58476d1ce4e5b9n) & MASK;
    z = ((z ^ (z >> 27n)) * 0x94d049bb133111ebn) & MASK;
    return (z ^ (z >> 31n)) & MASK;
  }

  double(): number {
    return Number(this.next() >> 11n) / 2 ** 53;
  }

  index(count: number): number {
    if (count <= 1) return 0;
    return Math.min(count - 1, Math.floor(this.double() * count));
  }
}

export interface SimulatedDay {
  offset: number;
  payments: number[];
  gross: number;
  paymentCount: number;
  /** Compratori distinti: meno dei pagamenti, perche' chi fa upsell paga due
   *  volte. */
  customers: number;
}

/** La giornata a `offset` giorni fa: 0 è oggi, 1 ieri. Il risultato dipende
 *  dalla data, non dall'offset: domani questa stessa giornata varra' ancora
 *  quanto vale adesso. */
/** La giornata si costruisce per clienti, non per pagamenti sciolti: ognuno
 *  prende il prodotto base, e una parte aggiunge quello piu' caro. */
export function day(simulation: Simulation, offset: number): SimulatedDay {
  const amounts = simulation.paymentAmounts.filter((a) => a > 0).sort((a, b) => a - b);
  if (amounts.length === 0) {
    return { offset, payments: [], gross: 0, paymentCount: 0, customers: 0 };
  }

  const base = amounts[0];
  const upsells = amounts.slice(1);
  const giorno = epochDay(offset);
  const rng = new SeededRandom((BigInt(simulation.seed) + BigInt(giorno) * 7919n) & MASK);

  const low = Math.min(simulation.dailyMin, simulation.dailyMax);
  const high = Math.max(simulation.dailyMin, simulation.dailyMax);

  // L'attivita' cresce: il livello sale dal minimo il giorno dell'apertura fino
  // al massimo oggi. Sopra ci sono due disturbi, il mese buono o scarso e lo
  // scarto del singolo giorno, che servono solo a non far venire una retta. Il
  // valore resta comunque dentro l'intervallo impostato.
  const arco = Math.max(1, Math.max(2, simulation.businessDays) - 1);
  const inizio = simulation.startDay ?? giorno + offset - arco;
  // La rampa e' piu' lunga dello storico che si vede: cosi' oggi sta all'80%
  // e non al tetto. Se arrivasse al massimo proprio oggi, le ultime settimane
  // verrebbero tutte schiacciate contro il tetto e piatte.
  const rampa = arco * 1.25;
  const progresso = Math.min(1, Math.max(0, (giorno - inizio) / rampa));

  const block = Math.floor(giorno / 30);
  const blockRng = new SeededRandom((BigInt(simulation.seed) + BigInt(block) * 2654435761n) & MASK);
  const onda = (blockRng.double() - 0.5) * 0.3;
  const spread = (rng.double() - 0.5) * 0.25;

  // Le estrazioni sopra avvengono comunque, anche quando oggi e' fissato:
  // saltarle sposterebbe il flusso del generatore e cambierebbe i giorni dopo.
  const casuale = low + Math.min(1, Math.max(0, 0.06 + 0.94 * progresso + onda + spread)) * (high - low);
  const fissato = simulation.dayTargets?.[dateKey(dateAt(offset))];
  const target = fissato && fissato > 0 ? fissato : casuale;

  // La quota di upsell del giorno, dal suo generatore: cosi' il flusso
  // principale resta identico fra TypeScript e Swift.
  const upsellRng = new SeededRandom((BigInt(simulation.seed) + BigInt(giorno) * 15485863n) & MASK);
  const quotaMin = Math.min(simulation.repeatMinPercent ?? 40, simulation.repeatMaxPercent ?? 80);
  const quotaMax = Math.max(simulation.repeatMinPercent ?? 40, simulation.repeatMaxPercent ?? 80);
  const quota = quotaMin + upsellRng.double() * (quotaMax - quotaMin);

  const payments: number[] = [];
  let customers = 0;
  let sum = 0;

  while (sum + base <= target && payments.length < 2000) {
    payments.push(base);
    sum += base;
    customers += 1;

    if (upsells.length === 0 || rng.double() * 100 >= quota) continue;
    const extra = upsells[rng.index(upsells.length)];
    if (sum + extra <= target) {
      payments.push(extra);
      sum += extra;
    }
  }

  return { offset, payments, gross: sum, paymentCount: payments.length, customers };
}

export function customerCount(_simulation: Simulation, d: SimulatedDay): number {
  return d.customers;
}

export function net(simulation: Simulation, gross: number): number {
  return gross * (1 - simulation.netDeductionPercent / 100);
}

// MARK: - Periodi

export const PERIODS = ['1W', '4W', '1Y', 'MTD', 'QTD', 'YTD', 'ALL'] as const;
export type Period = (typeof PERIODS)[number];

export function dayCount(period: Period, today: Date, businessDays: number): number {
  switch (period) {
    case '1W':
      return 7;
    case '4W':
      return 28;
    case '1Y':
      return 365;
    case 'MTD':
      return today.getDate();
    case 'QTD': {
      const quarterStartMonth = Math.floor(today.getMonth() / 3) * 3;
      const start = new Date(today.getFullYear(), quarterStartMonth, 1);
      return Math.round((today.getTime() - start.getTime()) / 86_400_000) + 1;
    }
    case 'YTD': {
      const start = new Date(today.getFullYear(), 0, 1);
      return Math.round((today.getTime() - start.getTime()) / 86_400_000) + 1;
    }
    case 'ALL':
      return Math.max(2, businessDays);
  }
}

/** Gli incassi giornalieri di un periodo, dal più vecchio a oggi. */
export function dailyGross(simulation: Simulation, from: number, to: number): number[] {
  const out: number[] = [];
  for (let offset = to; offset >= from; offset -= 1) {
    out.push(day(simulation, offset).gross);
  }
  return out;
}

/** I punti del grafico: un punto per giorno finché si leggono, altrimenti i
 *  giorni si sommano a gruppi. */
export function series(daily: number[], maxPoints = 14): number[] {
  if (daily.length <= maxPoints) return daily;

  // I gruppi si contano partendo da oggi e andando indietro: se il resto
  // finisse in fondo, l'ultimo gruppo avrebbe meno giorni degli altri e il
  // grafico crollerebbe a picco sull'ultimo punto senza che sia successo
  // niente. Il resto sta all'inizio, dove un valore più basso è plausibile.
  const bucket = Math.ceil(daily.length / maxPoints);
  const remainder = daily.length % bucket;

  const out: number[] = [];
  // Un avanzo troppo corto si unisce al gruppo dopo invece di fare punto a sé.
  let start = remainder >= bucket / 2 ? 0 : remainder;
  if (remainder > 0 && remainder >= bucket / 2) {
    out.push(daily.slice(0, remainder).reduce((a, b) => a + b, 0));
    start = remainder;
  }

  for (; start < daily.length; start += bucket) {
    out.push(daily.slice(start, start + bucket).reduce((a, b) => a + b, 0));
  }
  return out;
}

// MARK: - Riepiloghi per la pagina

export interface Totals {
  gross: number;
  net: number;
  payments: number;
  customers: number;
}

export function totals(simulation: Simulation, from: number, to: number): Totals {
  let gross = 0;
  let payments = 0;
  let customers = 0;

  for (let offset = from; offset <= to; offset += 1) {
    const d = day(simulation, offset);
    gross += d.gross;
    payments += d.paymentCount;
    customers += customerCount(simulation, d);
  }

  return { gross, net: net(simulation, gross), payments, customers };
}

/** Lo spaccato della card "Payments": quanto è andato a buon fine e quanto no. */
export function paymentBreakdown(simulation: Simulation, from: number, to: number) {
  const succeeded = totals(simulation, from, to).gross;
  return {
    succeeded,
    uncaptured: 0,
    refunded: succeeded * REFUNDED_RATIO,
    blocked: succeeded * BLOCKED_RATIO,
    failed: succeeded * FAILED_RATIO,
  };
}

/** L'incasso di oggi ora per ora, sommato: è la linea piena del grafico in
 *  alto. Si ferma all'ora corrente, perché la giornata non è finita. */
export function cumulativeByHour(
  simulation: Simulation,
  offset: number,
  upToHour: number,
): number[] {
  const d = day(simulation, offset);
  // Anche qui il seme e' il giorno assoluto: se fosse l'offset, a mezzanotte
  // le ore di ieri si ridistribuirebbero da capo.
  const rng = new SeededRandom((BigInt(simulation.seed) + BigInt(epochDay(offset)) * 104_729n) & MASK);

  // I pagamenti si distribuiscono sulle ore, con più peso nel pomeriggio.
  const hours = new Array(24).fill(0) as number[];
  for (const amount of d.payments) {
    const hour = Math.min(23, Math.floor(6 + rng.double() * 18));
    hours[hour] += amount;
  }

  const out: number[] = [];
  let running = 0;
  for (let hour = 0; hour <= Math.min(23, upToHour); hour += 1) {
    running += hours[hour];
    out.push(running);
  }
  return out;
}

export interface FailedPayment {
  amount: number;
  date: Date;
  id: string;
}

/** Le righe della card "Failed payments": importi veri presi dagli stessi
 *  tagli, con data e identificativo stabili. */
export function failedPayments(simulation: Simulation, count: number): FailedPayment[] {
  const amounts = simulation.paymentAmounts.filter((a) => a > 0).sort((a, b) => a - b);
  if (amounts.length === 0) return [];

  const rng = new SeededRandom((BigInt(simulation.seed) + 31n) & MASK);
  const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  const out: FailedPayment[] = [];
  let offset = 0;

  for (let i = 0; i < count; i += 1) {
    offset += 1 + rng.index(20);
    const date = new Date();
    date.setDate(date.getDate() - offset);
    date.setHours(rng.index(24), rng.index(60), 0, 0);

    let id = 'pi_3';
    for (let c = 0; c < 21; c += 1) id += alphabet[rng.index(alphabet.length)];

    out.push({ amount: amounts[rng.index(amounts.length)], date, id });
  }

  return out;
}

// MARK: - Formattazione

const SYMBOLS: Record<Currency, { card: string; report: string }> = {
  usd: { card: 'US$', report: 'US$' },
  eur: { card: '€', report: '€' },
  gbp: { card: '£', report: '£' },
  aed: { card: 'AED ', report: 'AED ' },
};

export function symbol(currency: Currency): string {
  return SYMBOLS[currency].report;
}

/** Formato americano a prescindere dalla lingua del browser: la dashboard di
 *  Stripe scrive 1,234.56 anche su un computer italiano. */
export function money(value: number, currency: Currency): string {
  return (
    symbol(currency) +
    value.toLocaleString('en-US', { minimumFractionDigits: 2, maximumFractionDigits: 2 })
  );
}

/** Le etichette dell'asse verticale: US$4k, US$3k… */
export function compactMoney(value: number, currency: Currency): string {
  const sym = symbol(currency);
  if (value >= 1000) return `${sym}${Math.round(value / 1000)}k`;
  return sym + Math.round(value);
}

export function isSimulation(value: unknown): value is Simulation {
  if (typeof value !== 'object' || value === null) return false;
  const s = value as Partial<Simulation>;
  return (
    typeof s.merchantName === 'string' &&
    typeof s.dailyMin === 'number' &&
    typeof s.dailyMax === 'number' &&
    Array.isArray(s.paymentAmounts) &&
    typeof s.seed === 'number'
  );
}
